src/index/CompressedBlockPrefiltering.cpp

```diff
@@ -0,0 +1,195 @@
+    case LE:
+      // Complement X <= Y: X > Y
+      return std::make_unique<GreaterThanExpression>(referenceId_);
+    case EQ:
+      // Complement X == Y: X != Y
+      return std::make_unique<NotEqualExpression>(referenceId_);
+    case NE:
+      // Complement X != Y: X == Y
+      return std::make_unique<EqualExpression>(referenceId_);
+    case GE:
+      // Complement X >= Y: X < Y
+      return std::make_unique<LessThanExpression>(referenceId_);
+    case GT:
+      // Complement X > Y: X <= Y
+      return std::make_unique<LessEqualExpression>(referenceId_);
+    default:
+      AD_FAIL();
+  }
+};
+
+//______________________________________________________________________________
+template <CompOp Comparison>
+std::vector<BlockMetadata> RelationalExpression<Comparison>::evaluate(
+    const std::vector<BlockMetadata>& input, size_t evaluationColumn) const {
+  using namespace valueIdComparators;
+  std::vector<ValueId> valueIdsInput;
+  // For each BlockMetadata value in vector input, we have a respective Id for
+  // firstTriple and lastTriple
+  valueIdsInput.reserve(2 * input.size());
+
+  for (const auto& block : input) {
+    valueIdsInput.push_back(
+        getIdFromColumnIndex(block.firstTriple_, evaluationColumn));
+    valueIdsInput.push_back(
+        getIdFromColumnIndex(block.lastTriple_, evaluationColumn));
+  }
+  // Use getRangesForId (from valueIdComparators) to extract the ranges
+  // containing the relevant ValueIds.
+  // For pre-filtering with CompOp::EQ, we have to consider empty ranges.
+  // Reason: The referenceId_ could be contained within the bounds formed by
+  // the IDs of firstTriple_ and lastTriple_ (set false flag to keep
+  // empty ranges).
+  auto relevantIdRanges =
+      Comparison != CompOp::EQ
+          ? getRangesForId(valueIdsInput.begin(), valueIdsInput.end(),
+                           referenceId_, Comparison)
+          : getRangesForId(valueIdsInput.begin(), valueIdsInput.end(),
+                           referenceId_, Comparison, false);
+  // The vector for relevant BlockMetadata values that contain ValueIds defined
+  // as relevant relevantIdRanges.
+  std::vector<BlockMetadata> relevantBlocks;
+  // Reserve memory, input.size() is upper bound.
+  relevantBlocks.reserve(input.size());
+  // Given the relevant Id ranges, retrieve the corresponding relevant
+  // BlockMetadata values from vector input and add them to the relevantBlocks
+  // vector.
+  for (const auto& [firstId, secondId] : relevantIdRanges) {
+    relevantBlocks.insert(
+        relevantBlocks.end(),
+        input.begin() + std::distance(valueIdsInput.begin(), firstId) / 2,
+        // Round up, for Ids contained within the bounding Ids of firstTriple
+        // and lastTriple we have to include the respective metadata block
+        // (that block is partially relevant).
+        input.begin() + std::distance(valueIdsInput.begin(), secondId + 1) / 2);
```

Comment:
technically you are not even allowed to write `secondId + 1` if `secondId == valueIdsInput.end()` but this probably works.
Better make an explicit if for this case.